platform/solr/solr-factory/src/main/java/org/codice/solr/client/solrj/SolrClient.java

```diff
@@ -0,0 +1,1030 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.solr.client.solrj;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.StreamingResponseCallback;
+import org.apache.solr.client.solrj.beans.DocumentObjectBinder;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.SolrPingResponse;
+import org.apache.solr.client.solrj.response.UpdateResponse;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.common.SolrDocumentList;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.util.NamedList;
+
+/**
+ * Provides a pure interface to the Solr client allowing us to add new functionality.
+ *
+ * <p><i>Note:</i> This is a precursor to having the whole Solr client abstracted into interfaces to
+ * decouple us from the Solr implementation.
+ */
+public interface SolrClient extends Closeable {
+  /**
+   * Retrieve the corresponding SolrJ's client.
+   *
+   * @return the corresponding SolrJ's client
+   */
+  public org.apache.solr.client.solrj.SolrClient getClient();
+
+  /**
+   * Gets the core associated with this Solr client
+   *
+   * @return the core associated with this Solr client
+   */
+  public String getCore();
+
+  /**
+   * Checks whether the Solr server this client is connected to and the corresponding core is
```

Comment:
❓I'm curious. Is there a reason why we don't want the `isAvailable()` and `isAvailable(long, TimeUnit)` to ping Solr right away instead of having all those warnings about cached information? From an interface usage perspective, it seems to make things more difficult than they should be. In other words, as a client I would expect to know the current state of the connection instead of the state at some unknown time. Starting a background poll and/or caching the result seem like implementation details.

Reply:
@lessarderic The main reason is that the ping() is there for exactly that and even after doing a ping, the server might still be down for the next request. Either way none of those are ever a guarantee.  Which is why I designed the isAvailable() to be right when it is false and be as best as it knows for true. 

@coyotesqrl I do not see the warning about the fact that returning true doesn't guarantee the next call to succeed as implementation details. It is a very big thing to be aware when using a client (even today) I would rather have that info slammed in a developer's face so that he is aware and doesn't blindly check if it is available before making a call which would yield nothing better.

I would agree about removing the indication about the background ping though.

⭕️

Reply:
👍

Reply:
👍